### PR TITLE
Registry center for Nacos

### PR DESCRIPTION
**Pull request by huangjian888**
Added Nacos Registry Module

### CODE REVIEW

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+    @Override
+    public String get(String key) {
+        return getDirectly(key);
+    }
+
+    @Override
+    public String getDirectly(String key) {
+        /**
+         * getConfig(String dataId, String group, long timeoutMs)
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java]: Can you remove inline comment? Only javadoc comment permitted.

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+import org.apache.shardingsphere.orchestration.reg.listener.DataChangedEvent;
+import org.apache.shardingsphere.orchestration.reg.listener.DataChangedEventListener;
+
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.Executor;
+
+@Slf4j
+public class NacosRegistryCenter implements RegistryCenter {
+    private ConfigService configService;
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java]: Need a blank line here, you may use checkstlye to check your code format

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+    @Override
+    public void persist(String key, String value) {
+        update(key,value);
+    }
+
+    @Override
+    public void update(String key, String value) {
+        /**
+         * publishConfig(String dataId, String group, String content)
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java]: Can you remove inline comment? Only javadoc comment permitted.

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+    @Override
+    public void persistEphemeral(String key, String value) {
+
+    }
+
+    @Override
+    public void watch(final String key, final DataChangedEventListener dataChangedEventListener) {
+        /**
+         * addListener(String dataId, String group, Listener listener)
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java]: Can you remove inline comment? Only javadoc comment permitted.

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+                @Override
+                public Executor getExecutor() {
+                    return null;
+                }
+
+                @Override
+                public void receiveConfigInfo(String configInfo) {
+                    /**
+                     *Nacos只有UPDATED状态
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java]: Can you remove inline comment? Only javadoc comment permitted.

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/test/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenterTest.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    @Test
+    public void assertUpdate() {
+        nacosRegistryCenter.update("sharding/test", "value2");
+    }
+
+    @Test
+    public void nacosAssertThat(){
+        assertThat(nacosRegistryCenter.get("sharding/test"), is("value1"));
+    }
+
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/test/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenterTest.java]: Do not need a blank line here

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/pom.xml [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.shardingsphere</groupId>
+            <artifactId>sharding-orchestration-reg-api</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.alibaba.nacos</groupId>
+            <artifactId>nacos-client</artifactId>
+            <version>${nacos-client.verison}</version>
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/pom.xml]: I checked source codes for nacos, there are still have some GPL license dependent. We should not include  nacos into a apache project, can you modify scope as `provided`?

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+         * dataId:
+         * 配置 ID，采用类似 package.class（如org.apache.shardingsphere）的命名规则保证全局唯一性，
+         * class 部分建议是配置的业务含义。 全部字符小写。只允许英文字符和 4 种特殊字符（"."、":"、"-"、"_"）。不超过 256 字节
+         * group:
+         * 配置分组，建议填写产品名：模块名（如 Nacos:Test）保证唯一性。 只允许英文字符和4种特殊字符（"."、":"、"-"、"_"），不超过128字节
+         * content:
+         * 配置内容，不超过 100K 字节
+         */
+        try {
+            String dataId = key.replace("/",".");
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java]: Here may break checkstyle, can you use checkstyle to check your codes first?

**Comment by huangjian888** [comment]
Thank you for your reply. I will modify the code according to your prompt, but it is not clear that the GPL license dependent, NACOS registry can not be integrated into shardingsphere？

**Comment by terrymanu** [comment]
Nacos-client dependent MySQL-JDBC driver which is GPL, so we can not inculde Nacos-client into ShardingSphere.
But we can let user to add Nacos-client by themselves, so it is fine if we change maven scope from `compile` (default) to `provided`.

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java [hunk 9]
```diff
@@ -58,21 +59,11 @@ public String get(String key) {
...
-         * dataId:
-         * 配置 ID，采用类似 package.class（如org.apache.shardingsphere）的命名规则保证全局唯一性，
-         * class 部分建议是配置的业务含义。 全部字符小写。只允许英文字符和 4 种特殊字符（"."、":"、"-"、"_"）。不超过 256 字节
-         * group:
-         * 配置分组，建议填写产品名：模块名（如 Nacos:Test）保证唯一性。 只允许英文字符和4种特殊字符（"."、":"、"-"、"_"），不超过128字节
-         * timeoutMs:
-         * 读取配置超时时间，单位 ms，推荐值 3000
-         */
         try {
             String dataId = key.replace("/",".");
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java]: This place still may break checkstyle

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java [hunk 10]
```diff
@@ -58,21 +59,11 @@ public String get(String key) {
 
     @Override
     public String getDirectly(String key) {
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java]: Input parameters should use final

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java [hunk 11]
```diff
@@ -122,18 +103,10 @@ public void persistEphemeral(String key, String value) {
...
-        /**
-         * addListener(String dataId, String group, Listener listener)
-         * dataId:
-         * 配置 ID，采用类似 package.class（如org.apache.shardingsphere）的命名规则保证全局唯一性，
-         * class 部分建议是配置的业务含义。 全部字符小写。只允许英文字符和 4 种特殊字符（"."、":"、"-"、"_"）。不超过 256 字节
-         * group:
-         * 配置分组，建议填写产品名：模块名（如 Nacos:Test）保证唯一性。 只允许英文字符和4种特殊字符（"."、":"、"-"、"_"），不超过128字节
-         */
         try {
             String dataId = key.replace("/",".");
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java]: This place still may break checkstyle

File: sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java [hunk 12]
```diff
@@ -122,18 +103,10 @@ public void persistEphemeral(String key, String value) {
...
             String group = properties.getProperty("group","SHARDING_SPHERE_DEFAULT_GROUP");
```
Comment (terrymanu) [inline sharding-orchestration/sharding-orchestration-reg/sharding-orchestration-reg-nacos/src/main/java/org/apache/shardingsphere/orchestration/reg/nacos/NacosRegistryCenter.java]: This place still may break checkstyle

**Comment by terrymanu** [comment]
Thank you, welcome be a contributor.
Wish your next contribution.

**Comment by terrymanu** [comment]
@huangjian888 
I just open a new issue for document of Nacos, you can have a look if you are interesting.

https://github.com/apache/incubator-shardingsphere-doc/issues/274

**Comment by dongzl** [comment]
Hi @terrymanu , I've always understood that shardingsphere only provide SPI and default implementation，for the other special implementation，implement shardingsphere‘s SPI，don't need mege to shardingsphere.

for example， for Registry center：
shardingsphere provide `org.apache.shardingsphere.orchestration.reg.api.RegistryCenter ` interface and default ZooKeeper Registry center, for the other special implementation, like Nacos、apollo，we can implement it and submit PR to shardingsphere?

**Comment by terrymanu** [comment]
@dongzl Sure, Other reg center which implement shardingsphere's SPI. And  we prefer accept those pull request into shardingsphere, and let user to choice by themselves.
If you are interesting for other reg center's implementations, you may create a new issue and provide pull request as same way.

**Comment by dongzl** [comment]
> @dongzl Sure, Other reg center which implement shardingsphere's SPI. And we prefer accept those pull request into shardingsphere, and let user to choice by themselves.
> If you are interesting for other reg center's implementations, you may create a new issue and provide pull request as same way.

OK, I understand, I'm very glad to do it.

**Comment by geomonlin** [comment]
Hi,Have you ever run proxy bootstrap to test Nacos?not through test, and I find that it is null pointer.Because regCenter.getChildrenKeys(configNode.getSchemaPath()) is null.
   @Override   public List<String> getChildrenKeys(final String key) { return null; }

**Comment by huangjian888** [comment]
> Hi,Have you ever run proxy bootstrap to test Nacos?not through test, and I find that it is null pointer.Because regCenter.getChildrenKeys(configNode.getSchemaPath()) is null.
> @override public List getChildrenKeys(final String key) { return null; }

At present, only JDBC mode is supported, but proxy mode is not yet adapted.
Proxy mode may need to be adapted by other developers

**Comment by tristaZero** [comment]
> > Hi,Have you ever run proxy bootstrap to test Nacos?not through test, and I find that it is null pointer.Because regCenter.getChildrenKeys(configNode.getSchemaPath()) is null.
> > @override public List getChildrenKeys(final String key) { return null; }
> 
> At present, only JDBC mode is supported, but proxy mode is not yet adapted.
> Proxy mode may need to be adapted by other developers

Hi, thanks for your contribution, our next release is coming, and your pr is a very important item in this release note. But we are sorry to see that it can not run well at proxy mode, which should be part of Nacos orchestration.

I wonder whether you are still interested in continuing doing this, and make this module become complete and perfect? 
There is no deadline, please at your own pace. If you want other developers to get on with it, can you create an issue to explain the task detail? Thanks a lot. :)

Moreover we should discuss how to put this contribution to this release, adding it with some remarks, or waiting it to be done in future.

**Comment by huangjian888** [comment]
I will submit the proxy mode pull request  as soon as possible

**Comment by tristaZero** [comment]
> I will submit the proxy mode pull request as soon as possible

Appreciated it very much, hope it could be at our release note and benefit ShardingSphere community. 👍

**Comment by tristaZero** [comment]
@geomonlin will help do some tests, please feel free to ask for help if your need. @huangjian888

**Comment by huangjian888** [comment]
> Hi,Have you ever run proxy bootstrap to test Nacos?not through test, and I find that it is null pointer.Because regCenter.getChildrenKeys(configNode.getSchemaPath()) is null.
> @override public List getChildrenKeys(final String key) { return null; }

I have submitted PR, you can see https://github.com/apache/initiator-shardingsphere/pull/3354

**Comment by huangjian888** [comment]
> font{ line-height: 1.6; } ul,ol{ padding-left: 20px; list-style-position: inside; } Now it can't be opened  https://github.com/apache/initiator-shardingsphere/pull/3354 lzs 1623537355@qq.com 签名由 网易邮箱大师 定制 On 10/24/2019 16:51，huangjian888<notifications@github.com> wrote： Hi,Have you ever run proxy bootstrap to test Nacos?not through test, and I find that it is null pointer.Because regCenter.getChildrenKeys(configNode.getSchemaPath()) is null. @override public List getChildrenKeys(final String key) { return null; } I have submitted PR, you can see https://github.com/apache/initiator-shardingsphere/pull/3354 —You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

you can see #3354

**Comment by geomonlin** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
    
        This is ok.I'm going to test it.
        
    
        
            
        
    
    
     
    
        
                    
            
                
                    
                            
                                
                            
                            
                                lzs
                            
                    
                        
                            
                                    1623537355@qq.com
                            
                        
                
            
        
        
    
    
        签名由
        网易邮箱大师
        定制
    
 

    On 10/24/2019 17:56，huangjian888<notifications@github.com> wrote： 




font{ line-height: 1.6; } ul,ol{ padding-left: 20px; list-style-position: inside; } Now it can't be opened  https://github.com/apache/initiator-shardingsphere/pull/3354 lzs 1623537355@qq.com 签名由 网易邮箱大师 定制 On 10/24/2019 16:51，huangjian888notifications@github.com wrote： Hi,Have you ever run proxy bootstrap to test Nacos?not through test, and I find that it is null pointer.Because regCenter.getChildrenKeys(configNode.getSchemaPath()) is null. @override public List getChildrenKeys(final String key) { return null; } I have submitted PR, you can see https://github.com/apache/initiator-shardingsphere/pull/3354 —You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.



you can see #3354



—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by geomonlin** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
    
        Hi.        NacosRegistryCenter.getDirectly()->configService.getConfig istill null 
    
        
            
        
    
    
     
    
        
                    
            
                
                    
                            
                                
                            
                            
                                lzs
                            
                    
                        
                            
                                    1623537355@qq.com
                            
                        
                
            
        
        
    
    
        签名由
        网易邮箱大师
        定制
    
 

    On 10/24/2019 17:56，huangjian888<notifications@github.com> wrote： 




font{ line-height: 1.6; } ul,ol{ padding-left: 20px; list-style-position: inside; } Now it can't be opened  https://github.com/apache/initiator-shardingsphere/pull/3354 lzs 1623537355@qq.com 签名由 网易邮箱大师 定制 On 10/24/2019 16:51，huangjian888notifications@github.com wrote： Hi,Have you ever run proxy bootstrap to test Nacos?not through test, and I find that it is null pointer.Because regCenter.getChildrenKeys(configNode.getSchemaPath()) is null. @override public List getChildrenKeys(final String key) { return null; } I have submitted PR, you can see https://github.com/apache/initiator-shardingsphere/pull/3354 —You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.



you can see #3354



—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by huangjian888** [comment]
You can see #3354  you need to configure the schema in the Nacos configuration center in advance, such as: nacosRegistryCenter.persist("/orchestration_ds/config/schema", "sharding_db");
then run proxy bootstrap. I've tested it

**Pull request merged**